cli/QuickstepCli.cpp

```diff
@@ -135,9 +141,22 @@ int main(int argc, char* argv[]) {
   gflags::ParseCommandLineFlags(&argc, &argv, true);
   google::InitGoogleLogging(argv[0]);
 
-  if (quickstep::FLAGS_num_workers > 0) {
-    printf("Starting Quickstep with %d worker thread(s)\n",
-           quickstep::FLAGS_num_workers);
+  // Detect the hardware concurrency level. Note this call will return 0
+  // if it fails (which it may on some machines/environments).
+  const unsigned int num_hw_threads = std::thread::hardware_concurrency();
+
+  // Use the command-line value if that was supplied, else use the value
+  // that we computed above, provided it did return a valid value.
+  // TODO(jmp): May need to change this at some point to keep one thread
+  //            available for the OS if the hardware concurrency level is high.
```

Comment:
std::thread::hardware_concurrency() returns the number of (virtual) CPU
cores on the host machine, not the maximum number of concurrent threads.
Right now, Foreman's CPU use is much lower than that of workers, so it's
not that much of a problem to create an extra foreman thread that can be
freely scheduled on any CPU. In the future, if we wind up running many
concurrent queries, it might make sense to reserve one or two cores
exclusively for foremen threads, but I don't think that's an issue for now.

On Thu, Jan 21, 2016 at 10:28 AM Zuyu ZHANG <notifications@github.com>
wrote:

> In cli/QuickstepCli.cpp
> <https://github.com/pivotalsoftware/quickstep/pull/7#discussion_r50441263>
> :
>
> > @@ -135,9 +141,22 @@ int main(int argc, char* argv[]) {
> >    gflags::ParseCommandLineFlags(&argc, &argv, true);
> >    google::InitGoogleLogging(argv[0]);
> >
> > -  if (quickstep::FLAGS_num_workers > 0) {
> > -    printf("Starting Quickstep with %d worker thread(s)\n",
> > -           quickstep::FLAGS_num_workers);
> > +  // Detect the hardware concurrency level. Note this call will return 0
> > +  // if it fails (which it may on some machines/environments).
> > +  const unsigned int num_hw_threads = std::thread::hardware_concurrency();
> > +
> > +  // Use the command-line value if that was supplied, else use the value
> > +  // that we computed above, provided it did return a valid value.
> > +  // TODO(jmp): May need to change this at some point to keep one thread
> > +  //            available for the OS if the hardware concurrency level is high.
>
> Do we need to reserve one hardware thread for Foreman? And one for
> Shiftboss in near future?
>
> FYI, Foreman will be changed as a long-running service to process
> concurrent query executions, instead of current one-query-per-Foreman model.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pivotalsoftware/quickstep/pull/7/files#r50441263>.
>